test/jdk/java/awt/regtesthelpers/PassFailJFrame.java

```diff
@@ -0,0 +1,148 @@
+     * @param maxStringLength number of columns of the instructional JTextArea
+     * @param timeoutMinutes  timeout of the test where time is specified in
+     *                        minutes.
+     * @throws HeadlessException
+     * @throws InterruptedException
+     * @throws InvocationTargetException
+     */
+    public PassFailJFrame(String title, String instructions,
+                          int maxRowLength, int maxStringLength,
+                          int timeoutMinutes) throws HeadlessException,
+            InterruptedException, InvocationTargetException {
+        super(title);
+        this.maxRowLength = maxRowLength;
+        this.maxStringLength = maxStringLength;
+        this.timeoutMinutes = timeoutMinutes;
+
+        invokeAndWait(() -> {
+            setLayout(new BorderLayout());
+            instructionsText = new JTextArea(instructions, maxRowLength, maxStringLength);
+            instructionsText.setEditable(false);
+            instructionsText.setFocusable(false);
+            add(instructionsText, BorderLayout.NORTH);
+
+            JButton btnPass = new JButton("Pass");
+            btnPass.addActionListener((e) -> latch.countDown());
+
+            JButton btnFail = new JButton("Fail");
+            btnFail.addActionListener((e) -> getFailureReason());
+
+            JPanel buttonsPanel = new JPanel();
+            buttonsPanel.add(btnPass);
+            buttonsPanel.add(btnFail);
+
+            add(buttonsPanel, BorderLayout.SOUTH);
+            pack();
+            setLocation(10, 10);
+            setVisible(true);
+        });
+    }
+
+    /**
+     * Wait for the user decision i,e user selects pass or fail button.
+     * If user does not select pass or fail button then the test waits for
+     * the specified timeoutMinutes period and the test gets timeout.
+     *
+     * @throws InterruptedException
+     * @throws InvocationTargetException
+     */
+    public void awaitAndCheck() throws InterruptedException, InvocationTargetException {
+        boolean timeoutHappened = !latch.await(this.timeoutMinutes,
+                TimeUnit.MINUTES);
+        if (isEventDispatchThread()) {
+            dispose();
+        } else invokeAndWait(() -> dispose());
+
+        if (timeoutHappened) {
+            throw new RuntimeException("Test timed out!");
+        }
+        if (failed) {
+            throw new RuntimeException("Test failed! : " + testFailedReason);
+        }
+    }
+
+    public void getFailureReason() {
+        final JDialog dialog = new JDialog();
+        dialog.setTitle("Failure reason");
+        JPanel jPanel = new JPanel(new BorderLayout());
+        JTextArea jTextArea = new JTextArea(5, 20);
+
+        JButton okButton = new JButton("Ok");
+        okButton.addActionListener((ae) -> {
+            testFailedReason = jTextArea.getText();
+            failed = true;
+            dialog.dispose();
+            latch.countDown();
+        });
```

Comment:
This doesn't handle closing the dialog via Close button or Esc. I believe it should also do everything but copying the entered text.

Reply:
fixed